generators/chipyard/src/main/scala/Cospike.scala

```diff
@@ -0,0 +1,108 @@
+package chipyard
+
+import chisel3._
+import chisel3.experimental.{IntParam, StringParam, IO}
+import chisel3.util._
+
+import freechips.rocketchip.config.{Parameters, Field, Config}
+import freechips.rocketchip.subsystem._
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.devices.tilelink._
+import freechips.rocketchip.util._
+
+import testchipip.TileTraceIO
+
+class CospikeResources(
+  isa: String,
+  pmpregions: Int,
+  mem0_base: BigInt,
+  mem0_size: BigInt,
+  nharts: Int,
+  bootrom: String
+) extends BlackBox(Map(
+  "ISA" -> StringParam(isa),
+  "PMPREGIONS" -> IntParam(pmpregions),
+  "MEM0_BASE" -> IntParam(mem0_base),
+  "MEM0_SIZE" -> IntParam(mem0_size),
+  "NHARTS" -> IntParam(nharts),
+  "BOOTROM" -> StringParam(bootrom)
+)) with HasBlackBoxResource {
+  val io = IO(new Bundle {})
+  addResource("/csrc/cospike.cc")
+  addResource("/vsrc/cospike.v")
+}
+
+case object SpikeCosimKey extends Field[Boolean](false)
+
+trait CanHaveSpikeCosim { this: ChipyardSystem =>
+  if (p(SpikeCosimKey)) {
+    InModuleBody {
+      val isa = tiles.headOption.map(_.isaDTS).getOrElse("")
+      val mem0_base = p(ExtMem).map(_.master.base).getOrElse(BigInt(0))
+      val mem0_size = p(ExtMem).map(_.master.size).getOrElse(BigInt(0))
+      val pmpregions = tiles.headOption.map(_.tileParams.core.nPMPs).getOrElse(0)
+      val nharts = tiles.size
+      val bootrom = bootROM.map(_.module.contents.toArray.mkString(" ")).getOrElse("")
+      val resources = Module(new CospikeResources(isa, pmpregions, mem0_base, mem0_size, nharts, bootrom))
+    }
+  }
+}
```

Comment:
I personally think that `EnableSpikeCosim` (i.e. setting the `SpikeCosimKey`) is a bit extra and unintuitive for people. Is there a nice way that you can think of where we can dump these parameters out always without the extra config fragment? Ideally I think it makes sense to just enable the traceio, then just connect up the harness binder and everything works.

I'm thinking of something like this:
- generate these parameters, dump them into a `.sv/.v` file as `\`defines` and have the `cospike.v` file read from those defines. 
- The `CospikeResources` wouldn't reference the `cospike.cc/.v` files (instead it would be its `defines.sv` that it generates.
- Then the actual `SpikeCosim` Chisel BB would be the only thing referencing `cospike.cc/.v` and would dump these files only when the harness binder is instantiated.
- The `cospike.v` `SpikeCosim` could have an `initial` block that set's up the `.cc` side of Spike so that you don't need the `cospike_resources` module.

Reply:
I've modified this so its just a HarnessBinder.